posts/MarkoB/2017-02-06-validator-service-loader.adoc

```diff
@@ -0,0 +1,310 @@
+users of your lib only with the relevant classes and hide implementation details from them, you don't want to
+expose your validator implementation by mentioning it as a `validatedBy` parameter of `@Constraint(validatedBy = { })`
+annotation (as well as making an import of it from your `*.impl` or `*.internal` package where all the magic things
+are happening).
+
+Let's get back to validating our beans - as that's why we all gathered here, right? :)
+
+Our new constraint annotation '@DurationMin' might look like this:
+
+```java
+@Documented
+@Constraint(validatedBy = { })
+@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE })
+@Retention(RUNTIME)
+@Repeatable(List.class)
+@ReportAsSingleViolation
+public @interface DurationMin {
+
+	String message() default "{com.acme.validation.constraints.DurationMin.message}";
+
+	Class<?>[] groups() default { };
+
+	Class<? extends Payload>[] payload() default { };
+
+	long value() default 0;
+
+	ChronoUnit units() default ChronoUnit.NANOS;
+
+	/**
+	 * Defines several {@code @DurationMin} annotations on the same element.
+	 */
+	@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE })
+	@Retention(RUNTIME)
+	@Documented
+	@interface List {
+
+		DurationMin[] value();
+	}
+}
+```
+
+And we would place it in a public package of our lib - `com.acme.validation.constraints`. Now that we
+have an annotation, we would want to start writing a validator for `Duration` and this newly created annotation.
+To do that you need to implement the `ConstraintValidator<DurationMin, Duration>` interface. An implementation
+might look like this:
+
+```java
+public class DurationMinValidator implements ConstraintValidator<DurationMin, Duration> {
+
+	private Duration duration;
+
+	@Override
+	public void initialize(DurationMin constraintAnnotation) {
+		this.duration = Duration.of( constraintAnnotation.value(), constraintAnnotation.units() );
+	}
+
+	@Override
+	public boolean isValid(Duration value, ConstraintValidatorContext context) {
+		// null values are valid
+		if ( value == null ) {
+			return true;
+		}
+		return duration.compareTo( value ) < 1;
+	}
+}
+```
+
+As we are creating a new constraint annotation, we should also provide a default message for it as well. This can be
+done by placing a `ContributorValidationMessages.properties` property file in the classpath. This property file should
+contain a key/message pair, where key is the one used in annotation declaration (in our case it's
+`com.acme.validation.constraints.DurationMin.message`) and message is the one you would like to show
+when validation fails. Our property file looks like:
+
+    com.acme.validation.constraints.DurationMin.message = must be greater than or equal to {value} {units}
+
+If you leave everything else as is, your constraint annotation will live its own life without knowing about the presence
```

Comment:
This could be condensed a bit. Just say that instead of referring from the constraint to the validator via `@Constraint#validatedBy()` for the reasons above we are using the service loader mechanism and show the file.